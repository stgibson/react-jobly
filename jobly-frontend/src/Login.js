import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import Container from "react-bootstrap/Container";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Button from "react-bootstrap/Button";
import Alert from "react-bootstrap/Alert";

/**
 * Component for displaying form for user to log in
 * @param {Object{function}} param0 
 * @returns JSX code for rendering login form
 */
const Login = ({ login }) => {
  const history = useHistory();
  const initFormData = { username: "", password: "" };
  const [formData, setFormData] = useState(initFormData);
  const [errors, setErrors] = useState([]);
  
  /**
   * Updates formData when user types in form
   * @param {Object{any}} evt 
   */
  const handleChange = evt => {
    const { id, value } = evt.target;
    setFormData(formData => ({ ...formData, [id]: value }));
  };

  /**
   * Logs in user and redirects to companies page
   * @param {Object{any}} evt 
   */
  const handleSubmit = async evt => {
    evt.preventDefault();
    const { errors } = await login(formData)
    if (errors.length) {
      setErrors(errors);
    }
    else {
      history.push("/companies");
    }
  };

  return (
    <Container>
      <div>
        <Row>
          <Col xs={ { span: 6, offset: 3 } }>
            <h3>Log In</h3>
            <Form onSubmit={ handleSubmit }>
              <Form.Group controlId="username">
                <Form.Label>Username</Form.Label>
                <Form.Control type="text" onChange={ handleChange } value={ formData.username } />
              </Form.Group>
              <Form.Group controlId="password">
                <Form.Label>Password</Form.Label>
                <Form.Control type="password" onChange={ handleChange }  value={ formData.password } />
              </Form.Group>
              { errors.map(error => <Alert variant="danger">{ error }</Alert>) }
              <Button variant="primary" type="submit">Submit</Button>
            </Form>
          </Col>
        </Row>
      </div>
    </Container>
  );
};

export default Login;