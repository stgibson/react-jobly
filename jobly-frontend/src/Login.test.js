import React from "react";
import { MemoryRouter } from "react-router-dom";
import { render, fireEvent } from "@testing-library/react";
import Login from "./Login";

const login = jest.fn(() => ({ errors: [] })); // Learned how to mock functions at https://jestjs.io/docs/mock-functions

it("renders without crashing", () => {
  render(<MemoryRouter><Login /></MemoryRouter>);
});

it("matches snapshot", () => {
  const { asFragment } = render(<MemoryRouter><Login /></MemoryRouter>);
  
  expect(asFragment()).toMatchSnapshot();
});

it("updates username input when user types", () => {
  const { queryByLabelText, queryByDisplayValue } = render(
    <MemoryRouter><Login /></MemoryRouter>
  );

  const input = queryByLabelText("Username");
  
  fireEvent.change(input, { target: { id: "username", value: "test" } });

  expect(queryByDisplayValue("test")).toBeInTheDocument();
});

it("updates password input when user types", () => {
  const { queryByLabelText, queryByDisplayValue } = render(
    <MemoryRouter><Login /></MemoryRouter>
  );

  const input = queryByLabelText("Password");
  
  fireEvent.change(input, { target: { id: "password", value: "password" } });

  expect(queryByDisplayValue("password")).toBeInTheDocument();
});

it("calls login when user submits form", () => {
  const { queryByLabelText, queryByText } = render(
    <MemoryRouter><Login login={ login } /></MemoryRouter>
  );

  const usernameInput = queryByLabelText("Username");
  const passwordInput = queryByLabelText("Password");
  const submitButton = queryByText("Submit");
  
  fireEvent.change(usernameInput, { target: { value: "test" } });
  fireEvent.change(passwordInput, { target: { value: "password" } });
  fireEvent.click(submitButton);

  expect(login.mock.calls.length).toEqual(1);
  expect(login.mock.calls[0][0].username).toEqual("test");
  expect(login.mock.calls[0][0].password).toEqual("password");
});
