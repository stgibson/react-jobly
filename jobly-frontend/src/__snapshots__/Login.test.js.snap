// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div>
      <div
        class="row"
      >
        <div
          class="col-6 offset-3"
        >
          <h3>
            Log In
          </h3>
          <form
            class=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                class="form-control"
                id="username"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                class="form-control"
                id="password"
                type="password"
                value=""
              />
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
